File: src/core/timeline_data.py
"""
Timeline data structures for NC-KTV
Manages multi-track timeline with clips, effects, and animation curves
"""

from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any, Literal
from enum import Enum
import json


class TrackType(Enum):
    """Types of timeline tracks"""
    AUDIO = "audio"
    VIDEO = "video"
    EFFECTS = "effects"
    LYRICS = "lyrics"


class EffectType(Enum):
    """Types of animation effects"""
    FADE_IN = "fade_in"
    FADE_OUT = "fade_out"
    SLIDE_LEFT = "slide_left"
    SLIDE_RIGHT = "slide_right"
    ZOOM_IN = "zoom_in"
    ZOOM_OUT = "zoom_out"
    BLUR = "blur"
    COLOR_SHIFT = "color_shift"
    CUSTOM = "custom"


class EasingCurve(Enum):
    """Animation easing curves"""
    LINEAR = "linear"
    EASE_IN = "ease_in"
    EASE_OUT = "ease_out"
    EASE_IN_OUT = "ease_in_out"
    CUSTOM_BEZIER = "custom_bezier"


@dataclass
class Effect:
    """Represents an animation effect with easing curve"""
    effect_type: EffectType
    start_time: float  # Seconds
    duration: float  # Seconds
    easing: EasingCurve = EasingCurve.LINEAR
    custom_curve_points: Optional[List[tuple]] = None  # Bezier control points [(x, y), ...]
    properties: Dict[str, Any] = field(default_factory=dict)  # Effect-specific properties
    
    @property
    def end_time(self) -> float:
        """Calculate end time of effect"""
        return self.start_time + self.duration
    
    def to_dict(self) -> dict:
        """Convert to dictionary"""
        return {
            'effect_type': self.effect_type.value,
            'start_time': self.start_time,
            'duration': self.duration,
            'easing': self.easing.value,
            'custom_curve_points': self.custom_curve_points,
            'properties': self.properties
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Effect':
        """Create from dictionary"""
        return cls(
            effect_type=EffectType(data['effect_type']),
            start_time=data['start_time'],
            duration=data['duration'],
            easing=EasingCurve(data.get('easing', 'linear')),
            custom_curve_points=data.get('custom_curve_points'),
            properties=data.get('properties', {})
        )


@dataclass
class Clip:
    """Represents a segment on a timeline track"""
    clip_id: str
    track_id: str
    start_time: float  # Position on timeline (seconds)
    duration: float  # Clip duration (seconds)
    source_file: Optional[str] = None  # Source file path for audio/video clips
    source_start: float = 0.0  # Start position in source file (for trimming)
    source_end: Optional[float] = None  # End position in source file
    effects: List[Effect] = field(default_factory=list)
    properties: Dict[str, Any] = field(default_factory=dict)  # Clip-specific properties
    
    @property
    def end_time(self) -> float:
        """Calculate end time on timeline"""
        return self.start_time + self.duration
    
    def move_to(self, new_start_time: float):
        """Move clip to new position on timeline"""
        if new_start_time < 0:
            new_start_time = 0.0
        self.start_time = new_start_time
    
    def resize(self, new_duration: float, from_start: bool = False):
        """
        Resize clip duration
        
        Args:
            new_duration: New duration in seconds
            from_start: If True, adjust start time instead of end time
        """
        MIN_DURATION = 0.1  # Minimum 100ms
        new_duration = max(MIN_DURATION, new_duration)
        
        if from_start:
            # Resizing from left edge - adjust start time and duration
            time_diff = self.duration - new_duration
            self.start_time += time_diff
            self.source_start += time_diff
        
        self.duration = new_duration
        
        # Update source_end if it exists
        if self.source_end is not None:
            self.source_end = self.source_start + new_duration
    
    def split_at(self, time: float) -> Optional['Clip']:
        """
        Split clip at specified time, returns new clip (right portion)
        
        Args:
            time: Time to split at (must be within clip bounds)
            
        Returns:
            New clip representing the right portion, or None if invalid
        """
        if time <= self.start_time or time >= self.end_time:
            return None
        
        # Calculate durations
        left_duration = time - self.start_time
        right_duration = self.end_time - time
        
        # Create right clip
        import uuid
        right_clip = Clip(
            clip_id=f"{self.clip_id}_split_{uuid.uuid4().hex[:8]}",
            track_id=self.track_id,
            start_time=time,
            duration=right_duration,
            source_file=self.source_file,
            source_start=self.source_start + left_duration,
            source_end=self.source_end,
            properties=self.properties.copy()
        )
        
        # Copy effects that occur in the right portion
        for effect in self.effects:
            effect_time = self.start_time + effect.start_time
            if effect_time >= time:
                # Effect is in right clip
                new_effect = Effect(
                    effect_type=effect.effect_type,
                    start_time=effect.start_time - left_duration,
                    duration=effect.duration,
                    easing=effect.easing,
                    custom_curve_points=effect.custom_curve_points,
                    properties=effect.properties.copy()
                )
                right_clip.effects.append(new_effect)
        
        # Adjust this (left) clip
        self.duration = left_duration
        if self.source_end is not None:
            self.source_end = self.source_start + left_duration
        
        # Remove effects that are now in right clip
        self.effects = [e for e in self.effects 
                       if (self.start_time + e.start_time) < time]
        
        return right_clip
    
    def add_effect(self, effect: Effect):
        """Add an effect to this clip"""
        self.effects.append(effect)
        # Sort effects by start time for consistent rendering
        self.effects.sort(key=lambda e: e.start_time)
    
    def remove_effect(self, effect_index: int) -> bool:
        """
        Remove effect by index
        
        Returns:
            True if effect was removed, False if index invalid
        """
        if 0 <= effect_index < len(self.effects):
            self.effects.pop(effect_index)
            return True
        return False
    
    def get_effect_at_time(self, time: float) -> List[Effect]:
        """Get all effects active at specified time (relative to clip start)"""
        return [e for e in self.effects 
                if e.start_time <= time <= e.end_time]
    
    def to_dict(self) -> dict:
        """Convert to dictionary"""
        return {
            'clip_id': self.clip_id,
            'track_id': self.track_id,
            'start_time': self.start_time,
            'duration': self.duration,
            'source_file': self.source_file,
            'source_start': self.source_start,
            'source_end': self.source_end,
            'effects': [e.to_dict() for e in self.effects],
            'properties': self.properties
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Clip':
        """Create from dictionary"""
        clip = cls(
            clip_id=data['clip_id'],
            track_id=data['track_id'],
            start_time=data['start_time'],
            duration=data['duration'],
            source_file=data.get('source_file'),
            source_start=data.get('source_start', 0.0),
            source_end=data.get('source_end'),
            properties=data.get('properties', {})
        )
        if 'effects' in data:
            clip.effects = [Effect.from_dict(e) for e in data['effects']]
        return clip


@dataclass
class Track:
    """Represents a timeline track that holds clips"""
    track_id: str
    track_type: TrackType
    name: str
    clips: List[Clip] = field(default_factory=list)
    is_muted: bool = False
    is_solo: bool = False
    is_locked: bool = False
    is_visible: bool = True
    height: int = 60  # Track height in pixels
    properties: Dict[str, Any] = field(default_factory=dict)
    
    def add_clip(self, clip: Clip):
        """Add a clip to this track"""
        clip.track_id = self.track_id
        self.clips.append(clip)
        # Sort clips by start time
        self.clips.sort(key=lambda c: c.start_time)
    
    def remove_clip(self, clip_id: str) -> bool:
        """Remove a clip by ID"""
        for i, clip in enumerate(self.clips):
            if clip.clip_id == clip_id:
                self.clips.pop(i)
                return True
        return False
    
    def get_clip(self, clip_id: str) -> Optional[Clip]:
        """Get clip by ID"""
        for clip in self.clips:
            if clip.clip_id == clip_id:
                return clip
        return None
    
    def get_clips_at_time(self, time: float) -> List[Clip]:
        """Get all clips active at a given time"""
        return [c for c in self.clips if c.start_time <= time <= c.end_time]
    
    def check_overlap(self, clip: Clip, exclude_clip_id: Optional[str] = None) -> bool:
        """
        Check if clip would overlap with any existing clips
        
        Args:
            clip: Clip to check
            exclude_clip_id: Clip ID to exclude from check (for moving existing clips)
            
        Returns:
            True if overlap detected, False otherwise
        """
        for existing_clip in self.clips:
            if exclude_clip_id and existing_clip.clip_id == exclude_clip_id:
                continue
            
            # Check for overlap
            if not (clip.end_time <= existing_clip.start_time or 
                   clip.start_time >= existing_clip.end_time):
                return True
        
        return False
    
    def find_next_free_slot(self, duration: float, after_time: float = 0.0) -> float:
        """
        Find the next available time slot for a clip of given duration
        
        Args:
            duration: Clip duration in seconds
            after_time: Start searching after this time
            
        Returns:
            Start time for the next free slot
        """
        candidate_time = after_time
        
        # Sort clips by start time
        sorted_clips = sorted(self.clips, key=lambda c: c.start_time)
        
        for clip in sorted_clips:
            if clip.start_time >= candidate_time:
                # Check if there's enough space before this clip
                if clip.start_time - candidate_time >= duration:
                    return candidate_time
                # Move past this clip
                candidate_time = clip.end_time
        
        # No clips or space after all clips
        return candidate_time
    
    def to_dict(self) -> dict:
        """Convert to dictionary"""
        return {
            'track_id': self.track_id,
            'track_type': self.track_type.value,
            'name': self.name,
            'clips': [c.to_dict() for c in self.clips],
            'is_muted': self.is_muted,
            'is_solo': self.is_solo,
            'is_locked': self.is_locked,
            'is_visible': self.is_visible,
            'height': self.height,
            'properties': self.properties
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Track':
        """Create from dictionary"""
        track = cls(
            track_id=data['track_id'],
            track_type=TrackType(data['track_type']),
            name=data['name'],
            is_muted=data.get('is_muted', False),
            is_solo=data.get('is_solo', False),
            is_locked=data.get('is_locked', False),
            is_visible=data.get('is_visible', True),
            height=data.get('height', 60),
            properties=data.get('properties', {})
        )
        if 'clips' in data:
            track.clips = [Clip.from_dict(c) for c in data['clips']]
        return track


class TimelineData:
    """Manages the complete timeline with all tracks"""
    
    def __init__(self):
        self.tracks: List[Track] = []
        self.duration: float = 0.0  # Total timeline duration
        self.playhead_position: float = 0.0
        self.zoom_level: float = 1.0  # Pixels per second
        self.snap_to_grid: bool = True
        self.grid_interval: float = 1.0  # Seconds
        self.metadata: Dict[str, Any] = {}
    
    def add_track(self, track: Track):
        """Add a track to the timeline"""
        self.tracks.append(track)
        self._update_duration()
    
    def remove_track(self, track_id: str) -> bool:
        """Remove a track by ID"""
        for i, track in enumerate(self.tracks):
            if track.track_id == track_id:
                self.tracks.pop(i)
                self._update_duration()
                return True
        return False
    
    def get_track(self, track_id: str) -> Optional[Track]:
        """Get track by ID"""
        for track in self.tracks:
            if track.track_id == track_id:
                return track
        return None
    
    def get_tracks_by_type(self, track_type: TrackType) -> List[Track]:
        """Get all tracks of a specific type"""
        return [t for t in self.tracks if t.track_type == track_type]
    
    def _update_duration(self):
        """Update timeline duration based on all clips"""
        max_end = 0.0
        for track in self.tracks:
            for clip in track.clips:
                max_end = max(max_end, clip.end_time)
        self.duration = max_end
    
    def snap_time(self, time: float) -> float:
        """Snap time to grid if enabled"""
        if not self.snap_to_grid:
            return time
        return round(time / self.grid_interval) * self.grid_interval
    
    def to_dict(self) -> dict:
        """Convert to dictionary"""
        return {
            'tracks': [t.to_dict() for t in self.tracks],
            'duration': self.duration,
            'playhead_position': self.playhead_position,
            'zoom_level': self.zoom_level,
            'snap_to_grid': self.snap_to_grid,
            'grid_interval': self.grid_interval,
            'metadata': self.metadata
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'TimelineData':
        """Create from dictionary"""
        timeline = cls()
        timeline.tracks = [Track.from_dict(t) for t in data.get('tracks', [])]
        timeline.duration = data.get('duration', 0.0)
        timeline.playhead_position = data.get('playhead_position', 0.0)
        timeline.zoom_level = data.get('zoom_level', 1.0)
        timeline.snap_to_grid = data.get('snap_to_grid', True)
        timeline.grid_interval = data.get('grid_interval', 1.0)
        timeline.metadata = data.get('metadata', {})
        return timeline
    
    def clear(self):
        """Clear all tracks and reset timeline"""
        self.tracks = []
        self.duration = 0.0
        self.playhead_position = 0.0

File: cpp/src/core/timeline/timeline_data.cpp
/*
 * NC-KTV Core — Timeline Data Implementation
 * Port of timeline_data.py (443 lines)
 */

#include "timeline_data.h"

#include <QUuid>
#include <algorithm>

namespace ncktv {

static std::string qs(const QString& s) { return s.toStdString(); }
static QString fromStd(const std::string& s) { return QString::fromStdString(s); }

// ─── Enum String Conversions ─────────────────────────────────────────────────

QString trackTypeToString(TrackType t) {
    switch (t) {
        case TrackType::Audio:   return "audio";
        case TrackType::Video:   return "video";
        case TrackType::Effects: return "effects";
        case TrackType::Lyrics:  return "lyrics";
    }
    return "audio";
}

TrackType trackTypeFromString(const QString& s) {
    if (s == "video")   return TrackType::Video;
    if (s == "effects") return TrackType::Effects;
    if (s == "lyrics")  return TrackType::Lyrics;
    return TrackType::Audio;
}

QString effectTypeToString(EffectType t) {
    switch (t) {
        case EffectType::FadeIn:     return "fade_in";
        case EffectType::FadeOut:    return "fade_out";
        case EffectType::SlideLeft:  return "slide_left";
        case EffectType::SlideRight: return "slide_right";
        case EffectType::ZoomIn:     return "zoom_in";
        case EffectType::ZoomOut:    return "zoom_out";
        case EffectType::Blur:       return "blur";
        case EffectType::ColorShift: return "color_shift";
        case EffectType::Custom:     return "custom";
    }
    return "fade_in";
}

EffectType effectTypeFromString(const QString& s) {
    if (s == "fade_out")    return EffectType::FadeOut;
    if (s == "slide_left")  return EffectType::SlideLeft;
    if (s == "slide_right") return EffectType::SlideRight;
    if (s == "zoom_in")     return EffectType::ZoomIn;
    if (s == "zoom_out")    return EffectType::ZoomOut;
    if (s == "blur")        return EffectType::Blur;
    if (s == "color_shift") return EffectType::ColorShift;
    if (s == "custom")      return EffectType::Custom;
    return EffectType::FadeIn;
}

QString easingCurveToString(EasingCurve c) {
    switch (c) {
        case EasingCurve::Linear:       return "linear";
        case EasingCurve::EaseIn:       return "ease_in";
        case EasingCurve::EaseOut:      return "ease_out";
        case EasingCurve::EaseInOut:    return "ease_in_out";
        case EasingCurve::CustomBezier: return "custom_bezier";
    }
    return "linear";
}

EasingCurve easingCurveFromString(const QString& s) {
    if (s == "ease_in")       return EasingCurve::EaseIn;
    if (s == "ease_out")      return EasingCurve::EaseOut;
    if (s == "ease_in_out")   return EasingCurve::EaseInOut;
    if (s == "custom_bezier") return EasingCurve::CustomBezier;
    return EasingCurve::Linear;
}

// ─── Effect JSON ─────────────────────────────────────────────────────────────

void to_json(nlohmann::json& j, const Effect& e) {
    j = {
        {"effect_type", qs(effectTypeToString(e.effectType))},
        {"start_time",  e.startTime},
        {"duration",    e.duration},
        {"easing",      qs(easingCurveToString(e.easing))},
        {"properties",  e.properties}
    };

    if (!e.customCurvePoints.isEmpty()) {
        auto pts = nlohmann::json::array();
        for (const auto& p : e.customCurvePoints)
            pts.push_back({p.first, p.second});
        j["custom_curve_points"] = pts;
    }
}

void from_json(const nlohmann::json& j, Effect& e) {
    e.effectType = effectTypeFromString(fromStd(j.at("effect_type").get<std::string>()));
    e.startTime  = j.at("start_time").get<double>();
    e.duration   = j.at("duration").get<double>();
    e.easing     = easingCurveFromString(fromStd(j.value("easing", "linear")));
    e.properties = j.value("properties", nlohmann::json::object());

    e.customCurvePoints.clear();
    if (j.contains("custom_curve_points")) {
        for (const auto& p : j["custom_curve_points"])
            e.customCurvePoints.append({p[0].get<double>(), p[1].get<double>()});
    }
}

// ─── Clip ────────────────────────────────────────────────────────────────────

void Clip::moveTo(double newStartTime) {
    startTime = std::max(0.0, newStartTime);
}

void Clip::resize(double newDuration, bool fromStart) {
    newDuration = std::max(0.01, newDuration);   // Minimum 10ms

    if (fromStart) {
        double oldEnd = endTime();
        double newStart = oldEnd - newDuration;
        if (newStart < 0.0) {
            newStart = 0.0;
            newDuration = oldEnd;
        }
        startTime = newStart;
    }
    duration = newDuration;
}

std::optional<Clip> Clip::splitAt(double time) {
    if (time <= startTime || time >= endTime())
        return std::nullopt;

    // Create right portion
    Clip right;
    right.clipId      = QUuid::createUuid().toString(QUuid::WithoutBraces);
    right.trackId     = trackId;
    right.startTime   = time;
    right.duration    = endTime() - time;
    right.sourceFile  = sourceFile;
    right.sourceStart = sourceStart + (time - startTime);
    right.sourceEnd   = sourceEnd;
    right.metadata    = metadata;

    // Adjust left portion (this clip)
    duration = time - startTime;

    // Split effects between left and right
    QVector<Effect> leftEffects, rightEffects;
    for (const auto& eff : effects) {
        double effAbsStart = startTime + eff.startTime;
        double effAbsEnd   = effAbsStart + eff.duration;

        if (effAbsEnd <= time) {
            leftEffects.append(eff);
        } else if (effAbsStart >= time) {
            Effect adjusted = eff;
            adjusted.startTime = eff.startTime - (time - startTime);
            rightEffects.append(adjusted);
        } else {
            // Effect spans the split — duplicate on both sides
            Effect leftEff  = eff;
            leftEff.duration = time - effAbsStart;
            leftEffects.append(leftEff);

            Effect rightEff  = eff;
            rightEff.startTime = 0.0;
            rightEff.duration  = effAbsEnd - time;
            rightEffects.append(rightEff);
        }
    }

    effects       = leftEffects;
    right.effects = rightEffects;

    return right;
}

void Clip::addEffect(const Effect& effect) {
    effects.append(effect);
}

bool Clip::removeEffect(int index) {
    if (index < 0 || index >= effects.size()) return false;
    effects.removeAt(index);
    return true;
}

QVector<const Effect*> Clip::getEffectsAtTime(double time) const {
    QVector<const Effect*> result;
    for (const auto& eff : effects) {
        if (eff.startTime <= time && time <= eff.endTime())
            result.append(&eff);
    }
    return result;
}

nlohmann::json Clip::toJson() const {
    nlohmann::json j = {
        {"clip_id",      qs(clipId)},
        {"track_id",     qs(trackId)},
        {"start_time",   startTime},
        {"duration",     duration},
        {"source_start", sourceStart},
        {"metadata",     metadata}
    };

    if (sourceFile.has_value())
        j["source_file"] = qs(sourceFile.value());
    if (sourceEnd.has_value())
        j["source_end"] = sourceEnd.value();

    auto effs = nlohmann::json::array();
    for (const auto& e : effects) {
        nlohmann::json ej;
        to_json(ej, e);
        effs.push_back(ej);
    }
    j["effects"] = effs;

    return j;
}

Clip Clip::fromJson(const nlohmann::json& j) {
    Clip c;
    c.clipId      = fromStd(j.at("clip_id").get<std::string>());
    c.trackId     = fromStd(j.at("track_id").get<std::string>());
    c.startTime   = j.at("start_time").get<double>();
    c.duration    = j.at("duration").get<double>();
    c.sourceStart = j.value("source_start", 0.0);
    c.metadata    = j.value("metadata", nlohmann::json::object());

    if (j.contains("source_file") && !j["source_file"].is_null())
        c.sourceFile = fromStd(j["source_file"].get<std::string>());
    if (j.contains("source_end") && !j["source_end"].is_null())
        c.sourceEnd = j["source_end"].get<double>();

    if (j.contains("effects")) {
        for (const auto& ej : j["effects"]) {
            Effect e;
            from_json(ej, e);
            c.effects.append(e);
        }
    }

    return c;
}

// ─── Track ───────────────────────────────────────────────────────────────────

void Track::addClip(const Clip& clip) {
    clips.append(clip);
    std::sort(clips.begin(), clips.end(),
              [](const Clip& a, const Clip& b) { return a.startTime < b.startTime; });
}

bool Track::removeClip(const QString& clipId) {
    auto it = std::find_if(clips.begin(), clips.end(),
                           [&](const Clip& c) { return c.clipId == clipId; });
    if (it == clips.end()) return false;
    clips.erase(it);
    return true;
}

Clip* Track::getClip(const QString& clipId) {
    for (auto& c : clips)
        if (c.clipId == clipId) return &c;
    return nullptr;
}

const Clip* Track::getClip(const QString& clipId) const {
    for (const auto& c : clips)
        if (c.clipId == clipId) return &c;
    return nullptr;
}

QVector<const Clip*> Track::getClipsAtTime(double time) const {
    QVector<const Clip*> result;
    for (const auto& c : clips) {
        if (c.startTime <= time && time < c.endTime())
            result.append(&c);
    }
    return result;
}

bool Track::checkOverlap(const Clip& clip, const QString& excludeClipId) const {
    for (const auto& existing : clips) {
        if (!excludeClipId.isEmpty() && existing.clipId == excludeClipId)
            continue;
        if (clip.startTime < existing.endTime() && clip.endTime() > existing.startTime)
            return true;
    }
    return false;
}

double Track::findNextFreeSlot(double duration, double afterTime) const {
    double candidate = afterTime;

    // Sort clips by start time (already sorted after addClip, but be safe)
    QVector<const Clip*> sorted;
    for (const auto& c : clips) sorted.append(&c);
    std::sort(sorted.begin(), sorted.end(),
              [](const Clip* a, const Clip* b) { return a->startTime < b->startTime; });

    for (const auto* c : sorted) {
        if (c->endTime() <= candidate) continue;   // clip is before our candidate
        if (candidate + duration <= c->startTime)
            return candidate;   // gap found
        candidate = c->endTime();   // skip past this clip
    }

    return candidate;
}

nlohmann::json Track::toJson() const {
    nlohmann::json j = {
        {"track_id",   qs(trackId)},
        {"track_type", qs(trackTypeToString(trackType))},
        {"name",       qs(name)},
        {"is_muted",   isMuted},
        {"is_solo",    isSolo},
        {"is_locked",  isLocked}
    };

    auto clipsArr = nlohmann::json::array();
    for (const auto& c : clips)
        clipsArr.push_back(c.toJson());
    j["clips"] = clipsArr;

    return j;
}

Track Track::fromJson(const nlohmann::json& j) {
    Track t;
    t.trackId   = fromStd(j.at("track_id").get<std::string>());
    t.trackType = trackTypeFromString(fromStd(j.at("track_type").get<std::string>()));
    t.name      = fromStd(j.at("name").get<std::string>());
    t.isMuted   = j.value("is_muted", false);
    t.isSolo    = j.value("is_solo", false);
    t.isLocked  = j.value("is_locked", false);

    if (j.contains("clips")) {
        for (const auto& cj : j["clips"])
            t.clips.append(Clip::fromJson(cj));
    }

    return t;
}

// ─── TimelineData ────────────────────────────────────────────────────────────

Track& TimelineData::addTrack(const QString& id, TrackType type, const QString& name) {
    Track t;
    t.trackId   = id;
    t.trackType = type;
    t.name      = name;
    tracks.append(t);
    return tracks.last();
}

bool TimelineData::removeTrack(const QString& trackId) {
    auto it = std::find_if(tracks.begin(), tracks.end(),
                           [&](const Track& t) { return t.trackId == trackId; });
    if (it == tracks.end()) return false;
    tracks.erase(it);
    return true;
}

Track* TimelineData::getTrack(const QString& trackId) {
    for (auto& t : tracks)
        if (t.trackId == trackId) return &t;
    return nullptr;
}

const Clip* TimelineData::getClipAtTime(double time) const {
    for (const auto& track : tracks) {
        if (track.isMuted) continue;
        auto clipsAtTime = track.getClipsAtTime(time);
        if (!clipsAtTime.isEmpty())
            return clipsAtTime.first();
    }
    return nullptr;
}

void TimelineData::recalculateDuration() {
    totalDuration = 0.0;
    for (const auto& track : tracks)
        for (const auto& clip : track.clips)
            totalDuration = std::max(totalDuration, clip.endTime());
}

nlohmann::json TimelineData::toJson() const {
    nlohmann::json j;
    j["total_duration"] = totalDuration;

    auto arr = nlohmann::json::array();
    for (const auto& t : tracks)
        arr.push_back(t.toJson());
    j["tracks"] = arr;

    return j;
}

TimelineData TimelineData::fromJson(const nlohmann::json& j) {
    TimelineData td;
    td.totalDuration = j.value("total_duration", 0.0);

    if (j.contains("tracks")) {
        for (const auto& tj : j["tracks"])
            td.tracks.append(Track::fromJson(tj));
    }

    return td;
}

} // namespace ncktv
